### PR TITLE
Add support for AWS GlobalAccelerator BYOIP.

### LINKED ISSUE #14002: Support BYOIP for Global Accelerator in AWS Provider
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

AWS added the ability for customers to provide their own IP addresses (BYOIP) when provisioning a Global Accelerator accelerator. This feature request is to add support for BYOIP i Global Accelerator in the Terraform AWS provider.

### New or Affected Resource(s)

`aws_globalaccelerator_accelerator`

### PR DESCRIPTION
**Pull request by coolfrood**
### Description
This PR adds support for BYOIP IP addresses for creating AWS GlobalAccelerator accelerators

### References
[API reference](https://docs.aws.amazon.com/global-accelerator/latest/api/API_CreateAccelerator.html#API_CreateAccelerator_RequestSyntax)

### Relations
Closes #14002

### Output from Acceptance Testing
(Changed IP address because of security reasons)

```
$ GLOBALACCELERATOR_BYOIP_IPV4_ADDRESS=127.0.0.1 make testacc TESTS=TestAccGlobalAcceleratorAccelerator_byoip PKG=globalaccelerator
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/globalaccelerator/... -v -count 1 -parallel 20 -run='TestAccGlobalAcceleratorAccelerator_byoip'  -timeout 180m
=== RUN   TestAccGlobalAcceleratorAccelerator_byoip
=== PAUSE TestAccGlobalAcceleratorAccelerator_byoip
=== CONT  TestAccGlobalAcceleratorAccelerator_byoip
--- PASS: TestAccGlobalAcceleratorAccelerator_byoip (59.49s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/globalaccelerator	59.571s

$ make testacc TESTS=TestAccGlobalAcceleratorAccelerator_byoip PKG=globalaccelerator
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/globalaccelerator/... -v -count 1 -parallel 20 -run='TestAccGlobalAcceleratorAccelerator_byoip'  -timeout 180m
=== RUN   TestAccGlobalAcceleratorAccelerator_byoip
=== PAUSE TestAccGlobalAcceleratorAccelerator_byoip
=== CONT  TestAccGlobalAcceleratorAccelerator_byoip
    accelerator_test.go:337: Environment variable GLOBALACCELERATOR_BYOIP_IPV4_ADDRESS not set
--- SKIP: TestAccGlobalAcceleratorAccelerator_byoip (0.53s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/globalaccelerator	0.604s
```

I am a member of AWS Global Accelerator engineering team.

### CODE REVIEW

File: internal/service/globalaccelerator/accelerator.go [hunk 1]
```diff
@@ -136,6 +142,10 @@ func resourceAcceleratorCreate(d *schema.ResourceData, meta interface{}) error {
 		input.IpAddressType = aws.String(v.(string))
 	}
 
+	if v, ok := d.GetOk("ip_addresses"); ok {
+		input.IpAddresses = expandIpAddresses(v.([]interface{}))
```
[the remark this answers is not in the document]
Comment (ewbankkit) [reply internal/service/globalaccelerator/accelerator.go:146]: Will need to add `d.Set("ip_addresses", aws.StringValueSlice(...))` to `resourceAcceleratorRead`.

File: internal/service/globalaccelerator/accelerator_test.go [hunk 2]
```diff
@@ -89,6 +91,54 @@ func TestAccGlobalAcceleratorAccelerator_ipAddressType_dualStack(t *testing.T) {
 	})
 }
 
+func TestAccGlobalAcceleratorAccelerator_byoip(t *testing.T) {
+	resourceName := "aws_globalaccelerator_accelerator.test"
+	rName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
+	requestedAddr := os.Getenv("GLOBALACCELERATOR_BYOIP_IPV4_ADDRESS")
```
Comment (ewbankkit) [inline internal/service/globalaccelerator/accelerator_test.go:97]: Please document in `docs/acc-test-environment-variables.md`.

**Comment by ewbankkit** [comment]
@coolfrood Thanks for the contribution 👏.
I've left some review comments.
Please also update the documentation page in `website/docs/r/globalaccelerator_accelerator.html.markdown`.

**Comment by coolfrood** [comment]
This @ewbankkit. I'll address your comments in next commit.

File: internal/service/globalaccelerator/accelerator.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (coolfrood) [reply internal/service/globalaccelerator/accelerator.go]: I thought about this. If you look at [CreateAccelerator API reference](https://docs.aws.amazon.com/global-accelerator/latest/api/API_CreateAccelerator.html), `IpAddresses` is an input parameter, whereas the output model has `IpSets`. I could technically extract the IP addresses from IP sets and put them into `ip_addresses`, but the output IP addresses are not the same is the input IP addresses; the input IP addresses are a subset of the output ones. When you read the resource, you don't know which ones were previously provided for input.

Any suggestions for how to deal with this?

**Comment by coolfrood** [comment]
@ewbankkit I have addressed all except one of your review comments.

File: internal/service/globalaccelerator/accelerator.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ewbankkit) [reply internal/service/globalaccelerator/accelerator.go]: I think we'll keep the top-level `ip_addresses` for BYOIP addresses only.
As you mention, on Read we can't know which addresses are BYOIP and which aren't so we can't set `ip_addresses`.
This means that the practitioner configured value will carry over (which is what we want).
The only side-effect of this is that the `ip_addresses` value won't be known during resource import.

**Comment by coolfrood** [comment]
Thanks for fixing the terrafmt error, @ewbankkit. Is there anything else you need?

**Comment by ewbankkit** [comment]
@coolfrood Thanks for the contribution :tada: :clap:.

**Pull request merged**